icefoganalytics/internal-data-portal
# Part 2: Data Owner: Edit an Existing Dataset

Partial completes https://github.com/icefoganalytics/internal-data-portal/issues/28

Relates to:
- https://github.com/icefoganalytics/internal-data-portal/issues/7
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)

# Context

## Dataset Description Manage Page - /dataset/:slug/description/manage

The edit version of the pages doesn’t seem to exist in the wireframes, so I would need another layer of User Role components to handle the different viewer types if I build this out. It might make more sense to just use auto-save for everything.


### Front-end routes

Various tabs on the same level as this component.


### Back-end routes



* /datasets/:id - GET 
* /datasets/:id - PATCH, update data set
* /tags - GET, list tags
* /tags - POST, create new tags
* /taggings - GET, list taggings
* /taggings - POST, associate existing tag with dataset
* /taggings/:id - DELETE, disassociate tag with data, orphaned tags should cleaned up
* /tags - POST, create new tag prior to associating it with a dataset via a tagging
* /users/search?email=...


### Components

* `BaseLayout.vue` - see previous definition
* `DatasetLayout.vue`
* `DatasetDescriptionManagePage.vue` - for data owner, sys admin and business analyst
    * `OwnerFormCard.vue` - may need to diverge from the wireframe slightly to support prefilling with edit after prefilling.
        * Owner Name - UserSelector.vue, Dataset#owner_id -> StewardshipEvolution#owner_id -> StewardshipEvolution#owner_name
        * Owner Position - pulled from owner position -> StewardshipEvolution#owner_position
        * Support Name - UserSelector.vue, StewardshipEvolution#support_id -> StewardshipEvolution#support_name
        * Support Email - pulled from the supporting user -> StewardshipEvolution#support_email
        * Support Position - pulled from the supporting user -> StewardshipEvolution#support_position
        * Department - pulled from owner department -> StewardshipEvolution#department
        * Division - pulled from owner division -> StewardshipEvolution#division
        * Branch - pulled from owner branch -> StewardshipEvolution#branch
        * Unit - pulled from owner unit -> StewardshipEvolution#unit

### Source Wireframe

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/68e8cbac-e574-4f8d-a048-631d877e8ecc)

# Implementation

Simplify data modeling, use DatasetStewardship in place of StewardshipEvolutions.
Relation is now one-to-one with Dataset, instead of one-to-many.
Rebuild dataset creation form.
Build Dataset manage form ownership sub-form.
Make QA Scenarios perform hard deletes.

# Screenshots

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/f936f9f3-7573-4901-a5e3-be5ea538b972)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. From the /dashboard page, create a new dataset via the create dataset form.
5. As you are the creator, you will be redirected to the datasets manage form immediately on creation.
6. Check that the dataset has all the fields you set during creation.
7. You can exit this via the breadcrumbs, go to the "Description" crumb to get to the not implemented "read" page.
8. Click on the "Datasets" crumb, to go to the table view.
9. Check that your new dataset loads correctly in the table.
10. You can click on it to go back to the "read" page.
11. Click the "Edit" button to go to the "manage" page.
12. Edit some of the "Owner" sub-form fields.
